solutions/usaco-718.mdx

```diff
@@ -2,29 +2,65 @@
 id: usaco-718
 source: USACO Gold 2017 February
 title: Why Did the Cow Cross The Road II
-author: Brad Ma
+author: Brad Ma, Jesse Choe
 ---
  
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_nocross_gold_feb17.html)
 
-The recursive approach where you try all configurations of crosswalks has the 
-time complexity $\mathcal{O}(2 ^ N)$, which is extremely slow. We observe that we have 
-many duplicate recursive calls where the configuration has been solved already, so by 
-using a top down dynamic progamming approach with memoization we can significantly reduce 
-the time complexity to $\mathcal{O}(N ^ 2)$. 
+For convenience, let's define $\texttt{dp}[i][j]$ represent the maximum number of disjoint friendly crosswalks up to field $i$ on the first side of the road and field $j$ on the second side of the road. Since the crosswalks must be disjoint, we can build a crosswalks as long as there isn't a crosswalk between fields $i$, $j - 1$ and $i - 1$, $j$. Thus, our transition is $dp[i][j] = max(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1] + (|id[i] - id[j]| <= 4))$, where $id[i]$ and $id[j]$ represent the breed IDs of the cows in the $i$-th and $j$-th fields, respectively.
 
-In this approach, we store our answers such that if there ever is a call to it again, 
-the function uses the cached answer instead of recomputing it. A bottom up dynamic 
-progamming approach without recursion can run faster because function calls have a 
-large constant, but it is more difficult to implement and is not necessary to pass all 
-test cases.
+The answer is $\texttt{dp}[N][N]$.
 
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N ^ 2)$
 
 <LanguageSection>
-  
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+using vl = vector<ll>;
+
+const ll MAXN = 1000;
```

Comment:
is there a reason `MAXN` is a long long?

Reply:
Just how I code (I don't use `int` anymore). I don't think this is too important.

Reply:
huh why would you use `long long` when you don't need to?? idk if this is bad practice but it doesn't seem good...

Reply:
hmm i think it's fine

Reply:
i agree w/ maggie, just change it to `int`